dali/python/nvidia/dali/plugin/pytorch/__init__.py

```diff
@@ -795,3 +853,266 @@ def __next__(self) -> List[Dict[str, torch.Tensor]]:
+        self.order_q = mp.Queue()
+        # Torch worker id, to be filled on first call to worker_id()
+        self._worker_id = None
+        # Iteration index for the current worker
+        self.data_idx = 0
+
+    def worker_id(self):
+        if self._worker_id is None:
+            self._worker_id = torch.utils.data.get_worker_info().id
+        return self._worker_id
+
+    class PipelineOutputRef:
+        """
+        Placeholder for a pipeline output reference, after the iteration has been scheduled to DALI.
+        """
+        def __init__(self, info):
+            self.info = info
+
+    def schedule_batch(self, inputs):
+        """
+        Schedule a pipeline run to DALI, by queuing the worker id, the iteration index and the inputs
+        """
+        # Identifier of this request
+        info = (self.worker_id(), self.data_idx)
+        with nvtx.annotate(f"dali_proxy.send_q.put {info}", color="blue"):
+            self.send_q.put((info, inputs))
+        self.data_idx = self.data_idx + 1
+        # Returns a placeholder, which is replaced with the actual data once the iteration completes
+        return DALIProxy.PipelineOutputRef(info)
+
+    class PipelineRunRef:
+        """
+        Placeholder for a pipeline run reference, which is returned by the data worker instead of the actual data
+
+        The PyTorch worker returns this trivial object, only containing information about this proxy instance and
+        the input data to the pipeline. Later in the collate function, we send the data for execution to DALI.
+        """
+        def __init__(self, dali_proxy, inputs):
+            self.dali_proxy = dali_proxy
+            self.inputs = inputs
+            assert len(self.inputs) == dali_proxy.num_inputs
+
+    def transform(self, *inputs):
+        """
+        The 'transform' function consists of returning a reference to the pipeline run
+        """
+        assert len(inputs) == self.num_inputs, f"Unexpected number of inputs: {len(inputs)}"
+        return DALIProxy.PipelineRunRef(self, inputs)
+
+
+def collate_pipeline_run_ref_fn(pipe_out, *, collate_fn_map=None):
+    """
+    Special collate function that schedules a batch for execution
+    """
+    assert len(pipe_out) > 0
+    first_elem = pipe_out[0]
+    inputs = [[] for idx in range(len(first_elem.inputs))]
+    for elem in pipe_out:
+        assert first_elem.dali_proxy == elem.dali_proxy
+        for idx, input_ref in enumerate(elem.inputs):
+            inputs[idx].append(input_ref)
+    return first_elem.dali_proxy.schedule_batch(inputs)
+
+
+def custom_collate(batch):
+    """
+    Subscribe a special collate function for PipelineRunRef, that handles the scheduling of the iteration
+    on the fly
+    """
+    collate_fn_map = data._utils.collate.default_collate_fn_map
+    collate_fn_map.update({DALIProxy.PipelineRunRef: collate_pipeline_run_ref_fn})
+    return collate(batch, collate_fn_map=collate_fn_map)
+
+
```

Comment:
Those are just internal helper functions, right? Maybe we should remove them from module scope (move inside a class in which they're used?) or at least prefix them with `_`?

Reply:
Mostly yes, but if someone is implementing a custom data loader they could reuse them as well